doc: add operator best practices (#4717)
operator-framework/operator-sdk: website/content/en/docs/best-practices/best-practices.md

weight: 1
description: This guide describes the best practices concepts to write operators.
---

## Development

Considerations for Operator developers:

- An Operator should manage a single type of application, essentially following the UNIX principle: do one thing and do it well.

- If an application consists of multiple tiers or components, multiple Operators should be written one for each of them. For example, if the application consists of Redis, AMQ and MySQL, there should be 3 Operators, not one.

- If there is significant orchestration and sequencing involved, an Operator should be written that represents the entire stack, in turn delegating to other Operators for orchestrating their part of it.

- Operators should own a CRD and only one Operator should control a CRD on a cluster. Two Operators managing the same CRD is not a recommended best practice. In the case where an API exists but with multiple implementations, this is typically an example of a no-op Operator because it doesn't have any deployment or reconciliation loop to define the shared API and other Operators depend on this Operator to provide one implementation of the API, e.g. similar to PVCs or Ingress.

> - Inside an Operator, multiple controllers should be used if multiple CRDs are managed. This helps with separation of concerns and code readability. Note that this doesn't necessarily mean one container image per controller, but rather one reconciliation loop (which could be running as part of the same Operator binary) per CRD.

Author: One controller per CRD and not 1 controller managing multi-apis.

- When defining CRDs, you should use OpenAPI spec to create a structural schema for your CRDs.

- Operators are instrumented to provide useful, actionable metrics to external systems (e.g. monitoring/alerting platforms).  Minimally, metrics should represent the software's health and key performance indicators, as well as support the creation of [service levels indicators](https://en.wikipedia.org/wiki/Service_level_indicator) such as throughput, latency, availability, errors, capacity, etc.

### Summary

- One Operator per managed application
- Multiple operators should be used for complex, multi-tier application stacks
- CRD can only be owned by a single Operator, shared CRDs should be owned by a separate Operator
- One controller per custom resource definition
- Use a tool like Operator SDK
- Do not hard-code namespaces or resources names
- Make watch namespace configurable
- Use semver / observe Kubernetes guidelines on versioning APIs
- Use OpenAPI spec with structural schema on CRDs
- Operators expose metrics to external systems

## Running On-Cluster

Considerations for on-cluster behavior

> - Like all containers on Kubernetes, Operators shouldn’t need to run as root unless absolutely necessary. Operators should come with their own ServiceAccount and not rely on the `default`.

Author: By default, the operators are now scaffolded with their own SA. 

> When you create a pod, if you do not specify a service account, it is automatically assigned the default service account in the same namespace

More info: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/

- Operators should not self-register their CRDs. These are global resources and careful consideration needs to be taken when setting those up. Also this requires the Operator to have global privileges which is potentially dangerous compared to that little extra convenience.

- Operators use CRs as the primary interface to the cluster user. As such, at all times, meaningful status information should be written to those objects unless they are solely used to store data in a structured schema.

> - Operators should update according to semver and should be updated frequently.

Reviewer: ```suggestion
- Operators should be updated frequently according to server versioning.
```

---
title: "Operator Best Practices"
linkTitle: "Best practices"
weight: 1
description: This guide describes the best practices concepts to write operators.
---

## Development

Considerations for operator developers:

- An Operator should manage a single type of application, essentially following the UNIX principle: do one thing and do it well.

> - If an application consists of multiple different tiers or components, multiple Operators should be written for each of them. If the application for example consists of Redis, AMQ and MySQL, there should be 3 Operators, not one.

Reviewer 2: ```suggestion
- If an application consists of multiple tiers or components, multiple Operators should be written one for each of them. For example, if the application consists of Redis, AMQ and MySQL, there should be 3 Operators, not one.
```

> - If there is significant orchestration and sequencing involved, an Operator should be written that represents the entire stack, in turn delegating to other Operators for orchestrating their part of it.

Reviewer 2: How would an operator delegate to another operator? Is this simply letting the other operator handle the reconciliation? This sentence is fine, but feels like we would have another item that talks about how the delegation would be handled.

Author: It came from https://github.com/operator-framework/community-operators/blob/master/docs/best-practices.md

I understand that it means that we would have for example Operator App and then another as DB. Then, we define that the APP depends on the DB. However, since it's very generic/conceptual I am not sure if we would able to add some concrete example.

Author: I think that is specified below either `An operator shouldn't deploy or manage other operators (such patterns are known as meta or super operators). It's the Operator Lifecycle Manager's job to manage the deployment and lifecycle of operators.` Note that we added a link for OLM. However, we can improve that in the future as well. Could that be?

> - Operators should own a CRD and only one Operator should control a CRD on cluster. Two Operators managing the same CRD calls for trouble. In case where an API exists but with multiple implementations, we typically have a no-op Operator (without any deployment/reconciliation loop) that defines that shared API and other Operators depend on this Operator to provide one implementation of the API, e.g. similar to PVCs or Ingress.

Author: ```suggestion
- Operators should own a CRD and only one Operator should control a CRD on a cluster. Two Operators managing the same CRD is not a recommended best practice. In the case where an API exists but with multiple implementations, this is typically an example of a no-op Operator because it doesn't have any deployment or reconciliation loop to define the shared API and other Operators depend on this Operator to provide one implementation of the API, e.g. similar to PVCs or Ingress.
```

- Inside an Operator, multiple controllers should be used if multiple CRDs are managed. This helps with separation of concerns and code readability. Note that this doesn't necessarily mean one container image per controller, but rather one reconciliation loop (which could be running as part of the same operator binary) per CRD.

> - An operator shouldn't deploy or manage other operators (such patterns are known as meta or super operators). It's the Operator Lifecycle Manager's job to manage the deployment and lifecycle of operators.

Reviewer 2: We probably need a document that describes how OLM solves this problem for these meta operators. Folks are creating these meta operators for a reason.

Reviewer 2: Also, we should either always capitalize Operator or not, seems we do a mixture in this document. Let's pick one and use it consistently, I prefer "Operator" :)

Author: > We probably need a document that describes how OLM solves this problem for these meta operators. Folks are creating these meta operators for a reason.

That is the OLM Dependency. I think we can link that here.

Author: Done.

> - If multiple operators must be packaged/shipped as a single entity (via the same CSV), then add all owned and required CRDs, as well as all deployments for operators that manage the owned CRDs, to the same CSV.

Author: ```suggestion
- If multiple operators are packaged and shipped as a single entity by the same CSV for example, then it is recommended to add all owned and required CRDs, as well as all deployments for operators that manage the owned CRDs, to the same CSV.
```

> - Writing an Operator involves using the Kubernetes API and there is always the same boilerplate code required to connect and interact with it. Use a framework like Operator-SDK to save yourself time with this and get a suite of tooling to ease development and testing.

Author: ```suggestion
- Writing an Operator involves using the Kubernetes API and as such it will always have the same boilerplate code required to connect and interact with it. Use a framework like the Operator-SDK to save yourself time with this and get a suite of tooling to ease development and testing.
```

> - Writing an Operator involves using the Kubernetes API and as such it will always have the same boilerplate code required to connect and interact with it. Use a framework like the Operator-SDK to save yourself time with this and get a suite of tooling to ease development and testing.

Reviewer 2: ```suggestion
- Writing an Operator involves using the Kubernetes API then it will always have the same boilerplate code required to connect and interact with it. Use a framework like the Operator SDK to save yourself time with this and get a suite of tooling to ease development and testing.
```

> - Operators shouldn’t make any assumptions about the namespace they are deployed in or hard-code names of resources that they expect to already exist.

Author: ```suggestion
- Operators shouldn’t make any assumptions about the namespace they are deployed in and they should not use hard-coded names of resources that they expect to already exist.
```

- Operators shouldn’t hard code the namespaces they are watching. This should be configurable - having no namespace supplied is interpreted as watching all namespaces

> - Semantic versioning (aka semver) should be used to version an Operator. Operators are long-running workloads on the cluster and it’s APIs are potentially in need of support over a longer period of time. Use the [semver.org guidelines](https://semver.org) to help determine when and how to bump versions when there are breaking or non-breaking changes.

Reviewer 2: ```suggestion
- Semantic versioning (aka semver) should be used to version an Operator. Operators are long-running workloads on the cluster and its APIs are potentially in need of support over a longer period of time. Use the [semver.org guidelines](https://semver.org) to help determine when and how to bump versions when there are breaking or non-breaking changes.
```

| **single version** | where the operator is tied to the version of the operand. |
| **hybrid approach** | where the operator is tied to a range of versions, and the user can select some level of the version. |

- An Operator should not deploy another Operator - an additional component on cluster should take care of this (OLM).

- When Operators change their APIs, CRD conversion (webhooks) should be used to deal with potentially older instances of them using the previous API version.

- Operators should make it easy for users to use their APIs - validating and rejecting malformed requests via extensive Open API validation schema on CRDs or via an admission webhook is good practice.

- The Operator itself should be really modest in its requirements - it should always be able to deploy by deploying its controllers, no user input should be required to start up the Operator.

- If user input is required to change the configuration of the Operator itself, a Configuration CRD should be used. Init-containers as part of the Operator deployments can be used to create a default instance of those CRs and then the Operator manages their lifecycle.

### Summary:

> An Operator...

Reviewer 2: ```suggestion
On the cluster, an Operator...
```

- Does not run as root

- Does not self-register CRDs

> - Does not install other Operators - rely on dependencies via package manager (OLM)

Reviewer 2: ```suggestion
- Does not install other Operators

- Does rely on dependencies via package manager (OLM)
```

---
title: "Operator Best Practices"
linkTitle: "Best practices"
weight: 1
description: This guide describes the best practices concepts to write operators.
---

# Operator Best Practices

## Development

Considerations for operator developers:

- An Operator should manage a single type of application, essentially following the UNIX principle: do one thing and do it well.

- If an application consists of multiple different tiers or components, multiple Operators should be written for each of them. If the application for example consists of Redis, AMQ and MySQL, there should be 3 Operators, not one.

- If there is significant orchestration and sequencing involved, an Operator should be written that represents the entire stack, in turn delegating to other Operators for orchestrating their part of it.

> - Operators should own a CRD and only one Operator should control a CRD on cluster. Two Operators managing the same CRD calls for trouble. In case where an API exists but with multiple implementations, we typically have a no-op Operator (without any deployment/reconciliation loop) that defines that shared API and other Operators depend on this Operator to provide one implementation of the API, e.g. similar to PVCs or Ingress.

Reviewer 3: Operators should own a CRD and only one Operator should control a CRD on a cluster. Two Operators managing the same CRD is not a recommended best practice. In the case where an API exists but with multiple implementations, this is typically an example of a no-op Operator becasue it doesn't have any deployment or reconciliation loop to define the shared API and other Operators depend on this Operator to provide one implementation of the API, e.g. similar to PVCs or Ingress.

- Inside an Operator, multiple controllers should be used if multiple CRDs are managed. This helps with separation of concerns and code readability. Note that this doesn't necessarily mean one container image per controller, but rather one reconciliation loop (which could be running as part of the same operator binary) per CRD.

- An operator shouldn't deploy or manage other operators (such patterns are known as meta or super operators). It's the Operator Lifecycle Manager's job to manage the deployment and lifecycle of operators.

> - If multiple operators must be packaged/shipped as a single entity (via the same CSV), then add all owned and required CRDs, as well as all deployments for operators that manage the owned CRDs, to the same CSV.

Reviewer 3: If multiple operators are packaged and shipped as a single entity by the same CSV for example, then it is recommended to add all owned and required CRDs, as well as all deployments for operators that manage the owned CRDs, to the same CSV.

> - Writing an Operator involves using the Kubernetes API and there is always the same boilerplate code required to connect and interact with it. Use a framework like Operator-SDK to save yourself time with this and get a suite of tooling to ease development and testing.

Reviewer 3: Writing an Operator involves using the Kubernetes API and as such it will always the same boilerplate code required to connect and interact with it. Use a framework like the Operator-SDK to save yourself time with this and get a suite of tooling to ease development and testing.

Reviewer 3: Writing an Operator involves using the Kubernetes API and as such it will always have the same boilerplate code required to connect and interact with it. Use a framework like the Operator-SDK to save yourself time with this and get a suite of tooling to ease development and testing.

> - Operators shouldn’t make any assumptions about the namespace they are deployed in or hard-code names of resources that they expect to already exist.

Reviewer 3: Operators shouldn’t make any assumptions about the namespace they are deployed in and they should not use hard-coded names of resources that they expect to already exist.
